RFC-001-libp2p.adoc

```diff
@@ -0,0 +1,361 @@
+
+Given that, we decided against gRPC.
+
+=== libp2p
+
+Libp2p is a modular network stack extracted from the ipfs (interplanetary file system) project.
+It provides reusable modules for common needs of peer-to-peer networking applications like discovery mechanisms, abstractions of various transport protocols, pluggable transport layer encryption and multiplexing.
+
+These are all features we find want for the transport layer of the COMIT network and hence we decided to build the communication layer on top of libp2p.
+
+== The COMIT messaging protocol
+
+=== Identification & configuration
+
+Application protocols in libp2p have an identifier which allows nodes to exchange, which protocols they understand.
+For COMIT and the messages described in this document, we use the protocol identifier: `/comit/1.0.0`.
+
+The libp2p stack of COMIT is configured as follows:
+
+- TCP transport
+- mplex multiplexer: https://github.com/libp2p/specs/tree/master/mplex/README.md
+- yamux multiplexer: https://github.com/hashicorp/yamux/blob/master/spec.md
+- secio encryption: https://github.com/libp2p/specs/blob/master/secio/README.md
+
+=== Versioning
+
+This document defines version `1.0.0` of the COMIT messaging protocol.
+
+=== Multiplexing
+
+Communication in a peer-to-peer network typically requires some form of multiplexing to allow every party to act as the initiator of a session if they want to actively send messages to another party.
+This section first briefly explains, how such multiplexing can be achieved if all you have are plain network sockets.
+As a second step, we will quickly cover, how the `substream` mechanism of libp2p affects this and thereby simplifies our application protocol.
+
+==== Multiplexing with plain sockets
+
+With traditional network sockets, there is always a listener and a dialer.
+The listening party has to have a port open the dialling party can connect to.
+If one party is behind a NAT, the only way to establish a network connection is for this party to dial to the other one.
+Establishing such a connection is expensive.
+We have to wait for the TCP handshake to complete and if there is transport encryption in place, keys might need to be exchanged before the first actual payload can be sent to the other party.
+Due to this cost and network limitations like NAT, it is good to reuse a connection between two parties once we have established one.
+This is where multiplexing comes into play.
+If both parties want to be able to send and receive messages at any point in time, those messages need to carry metadata so that the receiving end can correctly identify and associate messages.
+
+If all you have are plain network sockets, such a multiplexing strategy needs to be backed into your application protocol, which is a non-trivial increase in complexity.
+In fact, a former version of the COMIT messaging protocol (BAM!) had support for such functionality because we were not yet building on top of libp2p back then.
```

Comment:
What would be the value? The reason I included this whole section is because substreams are not a very common concept and I thought it is worth explaining this aspect.